packages/tx/src/types.ts

```diff
@@ -93,20 +95,91 @@ export function isAccessList(input: AccessListBytes | AccessList): input is Acce
 
 /**
  * Encompassing type for all transaction types.
- *
- * Note that this also includes legacy txs which are
- * referenced as {@link Transaction} for compatibility reasons.
  */
-export type TypedTransaction =
-  | Transaction
-  | AccessListEIP2930Transaction
-  | FeeMarketEIP1559Transaction
-  | BlobEIP4844Transaction
+export enum TransactionType {
+  Legacy,
+  AccessListEIP2930,
+  FeeMarketEIP1559,
+  BlobEIP4844,
+}
+
+export interface Transaction {
+  [TransactionType.Legacy]: LegacyTransaction
+  [TransactionType.FeeMarketEIP1559]: FeeMarketEIP1559Transaction
+  [TransactionType.AccessListEIP2930]: AccessListEIP2930Transaction
+  [TransactionType.BlobEIP4844]: BlobEIP4844Transaction
+}
+
+export type UnknownTransaction =
+  | Transaction[TransactionType.Legacy]
+  | Transaction[TransactionType.AccessListEIP2930]
+  | Transaction[TransactionType.FeeMarketEIP1559]
+  | Transaction[TransactionType.BlobEIP4844]
+
+export type TransactionsArray = Array<UnknownTransaction>
+
+export interface TransactionInterface<TTransactionType extends TransactionType> {
+  supports(capability: Capability): boolean
+  type: number
+  validate(): boolean
+  validate(stringError: false): boolean
+  validate(stringError: true): string[]
+  validate(stringError: boolean): boolean | string[]
+  getBaseFee(): bigint
+  getDataFee(): bigint
+  getUpfrontCost(): bigint
+  toCreationAddress(): boolean
+  raw(): TxValuesArray[TTransactionType]
+  serialize(): Uint8Array
+  getMessageToSign(hashMessage: false): Uint8Array | Uint8Array[]
+  getMessageToSign(hashMessage?: true): Uint8Array
+  hash(): Uint8Array
+  getMessageToVerifySignature(): Uint8Array
+  isSigned(): boolean
+  verifySignature(): boolean
+  getSenderAddress(): Address
+  getSenderPublicKey(): Uint8Array
+  sign(privateKey: Uint8Array): Transaction[TTransactionType]
+  toJSON(): JsonTx
+  errorStr(): string
+}
+
+export interface TxData {
+  [TransactionType.Legacy]: LegacyTxData
+  [TransactionType.AccessListEIP2930]: AccessListEIP2930TxData
+  [TransactionType.FeeMarketEIP1559]: FeeMarketEIP1559TxData
+  [TransactionType.BlobEIP4844]: BlobEIP4844TxData
+}
+
+export type UnknownTxData =
+  | TxData[TransactionType.Legacy]
+  | TxData[TransactionType.AccessListEIP2930]
+  | TxData[TransactionType.FeeMarketEIP1559]
+  | TxData[TransactionType.BlobEIP4844]
+
+export function isLegacyTxData(txData: UnknownTxData): txData is LegacyTxData {
+  const txType = Number(bytesToBigInt(toBytes(txData.type)))
+  return txType === 0
+}
+export function isAccessListEIP2930TxData(
+  txData: UnknownTxData
+): txData is AccessListEIP2930TxData {
+  const txType = Number(bytesToBigInt(toBytes(txData.type)))
+  return txType === 1
```

Comment:
Instead of 1,2,3 num comparisions, can do we enum `TransactionType` in a proper way assinging the type numbers to it explicitly and then use the enum for comparision.

Reply:
Good point, will do!

Reply:
Done! Also added additional typeguards and replaced a switch statement with that.